---
title: "ARG prediction benchmark analysis"
author: "Alberte"
date: "10/25/2024"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# ARG prediction benchmark analysis

```{r}
library(readxl)
library(tidyverse)
```


### Load combined arg prediction output

```{r}
arg_combined <- read_delim("/home/projects/cu_00014/people/albmol/healthypouch/HealthyPouch_article/output/arg_pred_sepseq_combined/combined_arg_pred_sepseq.txt", delim = "\t")

```


### Load metadata
```{r}
phenotype_metadata <- read_delim("/home/projects/cu_00014/data/sepseq_WGS/analysis/AMR_analysis/amr-pred-exp/data/df_ab_filtered_wide.tsv", delim = "\t") %>%  
  rename(sample_barcode = sample_id) %>%  
  filter(sample_barcode %in% arg_combined$sample_barcode)

```


### Benchmark seetings for accurate ARG prediction
```{r}

arg_combined_filtering <- arg_combined %>%  
  select(sample_barcode, Contig, Cut_Off, Pass_Bitscore, Best_Hit_Bitscore, Best_Hit_ARO, Best_Identities, ARO, Model_type, SNPs_in_Best_Hit_ARO, Other_SNPs, `Drug Class`, `Resistance Mechanism`, `AMR Gene Family`, 
         `Percentage Length of Reference Sequence`, Hit_Start, Hit_End, Antibiotic) %>%  
  filter(str_detect(Antibiotic, "ampicillin|cefotaxim|cefoxitin|ceftazidim|cefuroxim|ciprofloxacin|ertapenem|gentamicin|mecillinam|meropenem|piperacillin|sulfamethoxazole|trimethoprim|tobramycin")) %>% 
  mutate(type = "genotype") %>%  
  select(sample_barcode, Antibiotic, type)
  
  
  


metadata_long <- phenotype_metadata %>%  
  pivot_longer(
    cols = -sample_barcode,  
    names_to = "Antibiotic",  
    values_to = "Resistance") %>%
  filter(Resistance == "R") %>%
  group_by(sample_barcode) %>%  
  select(sample_barcode, Antibiotic) %>% 
  mutate(Antibiotic = tolower(Antibiotic)) %>%  
  mutate(type = "phenotype") %>%  
  mutate(Antibiotic = if_else("piptazo", "piperacillin", Antibiotic))


genotype_phenotype <- arg_combined_filtering %>%  
  bind_rows(metadata_long)



```


```{r}
ggplot(genotype_phenotype, aes(x = sample_barcode, y = Antibiotic)) + geom_point() + facet_grid(type ~ sample_barcode)



```


```{r}

filter()
ggplot(genotype_phenotype, aes(x = sample_barcode, y = Antibiotic)) + geom_point() + facet_wrap(sample_barcode ~ type)
```









